internal/tasks/image_gc.go

```diff
@@ -194,91 +200,102 @@ func (j *Janitor) executeGCPolicies(account keppel.Account, repo keppel.Reposito
 		}
 	}
 
-	//evaulate policies in order
+	//evaluate policies in order
 	proc := j.processor()
-	for _, p := range policies {
-		//for some time constraint matches, we need to know which manifests are
-		//still alive
-		var aliveManifests []keppel.Manifest
-		for _, m := range manifests {
-			if !m.IsDeleted {
-				aliveManifests = append(aliveManifests, m.Manifest)
-			}
+	for _, policy := range policies {
+		err := j.evaluatePolicy(proc, manifests, account, repo, policy)
+		if err != nil {
+			return err
 		}
+	}
 
-		//evaluate policy for each manifest
-		for _, m := range manifests {
-			//skip those manifests that are already deleted, and those which are
-			//protected by an earlier policy or one of the baseline checks above
-			if m.IsDeleted || m.GCStatus.IsProtected() {
-				continue
-			}
+	//finalize and persist GCStatus for all affected manifests
+	query = `UPDATE manifests SET gc_status_json = $1 WHERE repo_id = $2 AND digest = $3`
+	err = sqlext.WithPreparedStatement(j.db, query, manifests.persistGCStatus)
```

Comment:
The query string is tightly coupled to the behavior inside `persistGCStatus` and should be moved into that function. Also, I would suggest to have this be a function on `j`, in which case we would not need a type declaration for `manifestDataSlice`.
```suggestion
	err = j.persistGCStatus(manifests)
```

Reply:
The type is still used in evaluatePolicy()

Reply:
I don't think it is? evaluatePolicy takes a `[]*manifestData` instead of a `manifestDataSlice`.